src/name_utils.rs

```diff
@@ -0,0 +1,399 @@
+    #[strum(serialize = "RoleName")]
+    Role,
+    /// The CustomResource role group name
+    #[strum(serialize = "GroupName")]
+    Group,
+    /// The optional node name (e.g. for pods)
+    #[strum(serialize = "NodeName")]
+    Node,
+    /// Miscellaneous identifiers (e.g. "data" or "conf" for config maps)
+    #[strum(serialize = "MiscName")]
+    Misc,
+}
+
+/// Build a Kubernetes resource name. This is intended to work with the generatedName (add a unique
+/// hash) from Kubernetes. This method ensures that all single components do not exceed a certain
+/// length in order to keep the resource name below or equal to 63 (the maximum allowed characters)
+/// minus an offset for the unique hash (`KUBERNETES_HASH_MIN_LENGTH`).
+///
+/// In each sub name (`short_name`, `cluster_name`...) all non alphanumeric parts are automatically
+/// removed. After processing, each sub name is concatenated with "-" and added "-" as the last
+/// character (to separate from the kubernetes hash).
+///
+/// The basic sub name length for each sub name is calculated as follows:
+/// (1 / number_of_sub_names) * (RESOURCE_NAME_MAX_LEN - KUBERNETES_HASH_MIN_LENGTH - number_of_sub_names)
+///
+/// If sub names are shorter than the basic length, these characters can be used by another (longer)
+/// sub name.
+///
+/// # Arguments
+///
+/// * `short_name` - The short name of the custom resource definition.
+/// * `cluster_name` - The name of the custom resource.
+/// * `role_name` - The role name of the custom resource.
+/// * `group_name` - Optional group name of the custom resource.
+/// * `node_name` - Optional node name if available (e.g. for pods).
+/// * `misc_name` - Optional miscellaneous identifiers (e.g. "data" for config maps).
+///
+pub fn build_resource_name(
+    short_name: &str,
+    cluster_name: &str,
+    role_name: &str,
+    group_name: Option<&str>,
+    node_name: Option<&str>,
+    misc_name: Option<&str>,
+) -> OperatorResult<String> {
+    // collect a vector of all sub names and their respective kind
+    let mut sub_names = vec![
+        // transform to alphanumeric and lowercase
+        to_alphanumeric_lowercase_not_empty(SubName::Short, short_name)?,
+        to_alphanumeric_lowercase_not_empty(SubName::Cluster, cluster_name)?,
+        to_alphanumeric_lowercase_not_empty(SubName::Role, role_name)?,
+    ];
+
+    if let Some(group) = group_name {
+        sub_names.push(to_alphanumeric_lowercase_not_empty(SubName::Group, group)?);
+    }
+
+    if let Some(node) = node_name {
+        sub_names.push(to_alphanumeric_lowercase_not_empty(SubName::Node, node)?);
+    };
+
+    if let Some(misc) = misc_name {
+        sub_names.push(to_alphanumeric_lowercase_not_empty(SubName::Misc, misc)?);
+    };
+
+    // Calculate the maximum number of available characters
+    // The sub names vector length equals the size of additional required dashes
+    let max_chars = RESOURCE_NAME_MAX_LEN - KUBERNETES_HASH_MIN_LENGTH - sub_names.len();
```

Comment:
You can add one (+1) to this if you remove the trailing '-' that is always added now.

Reply:
It will stay (check above).